File: BOJ/2263.cpp
#include <iostream>
#include <stack>
#include <vector>

using namespace std;

struct Range {
	int in_order_start;
	int in_order_end;
	int post_order_start;
	int post_order_end;
};

int main(void) {
	int n; // the number of numbers

	vector<int>
		in_order_index, // index 0 not used
		post_order;

	stack<Range> printing_stack;

	ios::sync_with_stdio(false);
	cin.tie(NULL);

	// init
	cin >> n;
	in_order_index.resize(n + 1);
	post_order.resize(n);

	// get in-order
	for (int i = 0; i < n; i++) {
		int num;
		cin >> num;
		in_order_index[num] = i;
	}

	// get post-order
	for (int i = 0; i < n; i++) {
		cin >> post_order[i];
	}

	printing_stack.push({ 0, n - 1, 0, n - 1 });
	while (!printing_stack.empty()) {
		// pop one
		Range &range = printing_stack.top();
		int in_order_start = range.in_order_start;
		int in_order_end = range.in_order_end;
		int post_order_start = range.post_order_start;
		int post_order_end = range.post_order_end;
		printing_stack.pop();

		int root = post_order[post_order_end];
		int in_order_index_of_root = in_order_index[root];
		int n_left = in_order_index_of_root - in_order_start;
		int n_right = in_order_end - in_order_index_of_root;

		// print the local root
		cout << root << ' ';

		 // push all rights
		if (n_right > 0) {
			printing_stack.push({
				in_order_index_of_root + 1,
				in_order_end,
				post_order_start + n_left,
				post_order_end - 1,
			});
		}
		// push all lefts
		if (n_left > 0) {
			printing_stack.push({
				in_order_start,
				in_order_index_of_root - 1,
				post_order_start,
				post_order_start + n_left - 1,
			});
		}
	}

	return 0;
}

File: BOJ/2263.ts
import * as fs from 'fs';

const inputManager: Generator<string, any, unknown> = (function *() {
  const stdin = '/dev/stdin';
  const lines = fs.readFileSync(stdin).toString().split('\n');
  for (const line of lines) yield line;
  return null;
})();
const nextLine = (): string => inputManager.next().value;
const nextLineNumbers = () => nextLine().split(' ').map(Number);

// 구간 타입
type Range = {
  inOrderStart: number;
  inOrderEnd: number;
  postOrderStart: number;
  postOrderEnd: number;
};

// init
const [N] = nextLineNumbers();
const inOrder = nextLineNumbers();
const postOrder = nextLineNumbers();
const preOrder: Array<number> = [];

const inOrderIndex: Array<number> = [];
inOrder.forEach((value, index) => inOrderIndex[value] = index);

const printingStack: Array<Range> = [{
  inOrderStart: 0,
  inOrderEnd: N - 1,
  postOrderStart: 0,
  postOrderEnd: N - 1,
}];
while (printingStack.length) {
  // pop one
  const {
    inOrderStart,
    inOrderEnd,
    postOrderStart,
    postOrderEnd,
  } = printingStack.pop()!;
  const root = postOrder[postOrderEnd];
  const inOrderIndexOfRoot = inOrderIndex[root];
  const nLeft = inOrderIndexOfRoot - inOrderStart;
  const nRight = inOrderEnd - inOrderIndexOfRoot;

  // print the local root
  preOrder.push(root);

  // push all rights
  if (nRight > 0) {
    printingStack.push({
      inOrderStart: inOrderIndexOfRoot + 1,
      inOrderEnd: inOrderEnd,
      postOrderStart: postOrderStart + nLeft,
      postOrderEnd: postOrderEnd - 1,
    });
  }
  // push all lefts
  if (nLeft > 0) {
    printingStack.push({
      inOrderStart: inOrderStart,
      inOrderEnd: inOrderIndexOfRoot - 1,
      postOrderStart: postOrderStart,
      postOrderEnd: postOrderStart + nLeft - 1,
    });
  }
}

// print the pre-order result
console.log(preOrder.join(' '));
